File: modules/react-native-md-list/android/src/main/java/com/mdlist/md/MdModel.kt
package com.mdlist.md

/**
 * Flat, immutable representation of a parsed markdown document.
 *
 * A message is *flattened into rows* (one row = one block) instead of being kept
 * as a tree. That is the single most important decision for scrolling: a 4000
 * word assistant answer becomes ~60 small recyclable rows instead of one giant
 * cell, so RecyclerView/UITableView only ever measures + draws what is on screen.
 */

const val SPAN_BOLD = 1 shl 0
const val SPAN_ITALIC = 1 shl 1
const val SPAN_CODE = 1 shl 2
const val SPAN_STRIKE = 1 shl 3

class InlineSpan(
    @JvmField val start: Int,
    @JvmField val end: Int,
    @JvmField val style: Int,
    @JvmField val link: String?,
)

class InlineText(@JvmField val text: String, @JvmField val spans: List<InlineSpan>) {
    companion object {
        val EMPTY = InlineText("", emptyList())
    }
}

enum class RowType {
    /** assistant paragraph */
    PARAGRAPH,
    HEADING,
    LIST_ITEM,
    CODE,
    TABLE,
    DIVIDER,
    /** whole user message rendered as a right aligned bubble */
    BUBBLE,
}

const val ALIGN_LEFT = 0
const val ALIGN_CENTER = 1
const val ALIGN_RIGHT = 2

class MdTable(
    @JvmField val header: List<InlineText>,
    @JvmField val rows: List<List<InlineText>>,
    @JvmField val aligns: IntArray,
)

class MdRow(
    @JvmField val id: String,
    @JvmField val messageId: String,
    @JvmField val type: RowType,
    @JvmField val isUser: Boolean,
    @JvmField val inline: InlineText = InlineText.EMPTY,
    @JvmField val headingLevel: Int = 0,
    @JvmField val listMarker: String? = null,
    @JvmField val listDepth: Int = 0,
    @JvmField val quoteDepth: Int = 0,
    /** -1 = not a task item, 0 = unchecked, 1 = checked */
    @JvmField val checked: Int = -1,
    @JvmField val code: String? = null,
    @JvmField val language: String? = null,
    @JvmField val table: MdTable? = null,
    @JvmField val isFirstInMessage: Boolean = false,
    @JvmField val isLastInMessage: Boolean = false,
    /** true while the message is still streaming: its last row must never be cached */
    @JvmField val streaming: Boolean = false,
) {
    /** Cheap content fingerprint used by DiffUtil so unchanged rows are never re-bound. */
    @JvmField
    val contentHash: Int = run {
        var h = type.ordinal
        h = 31 * h + inline.text.hashCode()
        h = 31 * h + headingLevel
        h = 31 * h + (listMarker?.hashCode() ?: 0)
        h = 31 * h + listDepth
        h = 31 * h + quoteDepth
        h = 31 * h + checked
        h = 31 * h + (code?.hashCode() ?: 0)
        h = 31 * h + (language?.hashCode() ?: 0)
        h = 31 * h + (table?.let { t -> t.rows.size * 31 + t.header.size } ?: 0)
        h = 31 * h + (if (isFirstInMessage) 1 else 0)
        h = 31 * h + (if (isLastInMessage) 2 else 0)
        h
    }

    /** Layout cache key: identity + content, so a streamed edit invalidates only its own row. */
    @JvmField
    val layoutKey: String = "$id#$contentHash"
}

File: modules/react-native-md-list/ios/MdModel.swift
import Foundation

/// Mirror of the Kotlin model: a message is flattened into small, recyclable rows.
/// Keeping both platforms on the same data shape means one mental model, one set
/// of edge cases, and identical rendering rules.

struct MdInlineStyle: OptionSet {
    let rawValue: Int
    static let bold = MdInlineStyle(rawValue: 1 << 0)
    static let italic = MdInlineStyle(rawValue: 1 << 1)
    static let code = MdInlineStyle(rawValue: 1 << 2)
    static let strike = MdInlineStyle(rawValue: 1 << 3)
}

struct MdSpan {
    let start: Int   // UTF-16 offset
    let end: Int     // UTF-16 offset
    let style: MdInlineStyle
    let link: String?
}

struct MdInlineText {
    let text: String
    let spans: [MdSpan]
    static let empty = MdInlineText(text: "", spans: [])
}

enum MdRowType: Int {
    case paragraph = 0
    case heading = 1
    case listItem = 2
    case code = 3
    case table = 4
    case divider = 5
    case bubble = 6
}

enum MdAlign: Int {
    case left = 0, center = 1, right = 2
}

struct MdTable {
    let header: [MdInlineText]
    let rows: [[MdInlineText]]
    let aligns: [MdAlign]
}

final class MdRow {
    let id: String
    let messageId: String
    let type: MdRowType
    let inline: MdInlineText
    let headingLevel: Int
    let listMarker: String?
    let listDepth: Int
    let quoteDepth: Int
    /// -1 none, 0 unchecked, 1 checked
    let checked: Int
    let code: String?
    let language: String?
    let table: MdTable?
    var isFirstInMessage: Bool
    var isLastInMessage: Bool
    let streaming: Bool

    /// Content fingerprint: drives both diffing and the layout cache key.
    private(set) lazy var contentHash: Int = {
        var hasher = Hasher()
        hasher.combine(type.rawValue)
        hasher.combine(inline.text)
        hasher.combine(headingLevel)
        hasher.combine(listMarker)
        hasher.combine(listDepth)
        hasher.combine(quoteDepth)
        hasher.combine(checked)
        hasher.combine(code)
        hasher.combine(language)
        hasher.combine(table?.rows.count ?? 0)
        hasher.combine(isFirstInMessage)
        hasher.combine(isLastInMessage)
        return hasher.finalize()
    }()

    var layoutKey: String { "\(id)#\(contentHash)" }

    init(
        id: String,
        messageId: String,
        type: MdRowType,
        inline: MdInlineText = .empty,
        headingLevel: Int = 0,
        listMarker: String? = nil,
        listDepth: Int = 0,
        quoteDepth: Int = 0,
        checked: Int = -1,
        code: String? = nil,
        language: String? = nil,
        table: MdTable? = nil,
        isFirstInMessage: Bool = false,
        isLastInMessage: Bool = false,
        streaming: Bool = false
    ) {
        self.id = id
        self.messageId = messageId
        self.type = type
        self.inline = inline
        self.headingLevel = headingLevel
        self.listMarker = listMarker
        self.listDepth = listDepth
        self.quoteDepth = quoteDepth
        self.checked = checked
        self.code = code
        self.language = language
        self.table = table
        self.isFirstInMessage = isFirstInMessage
        self.isLastInMessage = isLastInMessage
        self.streaming = streaming
    }
}
